proposals/nnnn-mobile-dynamic-menu-cell-updating.md

```diff
@@ -0,0 +1,79 @@
+# Mobile Manager Dynamic Menu Cell Updating
+
+* Proposal: [SDL-NNNN](NNNN-mobile-dynamic-menu-cell-updating.md)
+* Author: [Joel Fischer](https://github.com/joeljfischer)
+* Status: **Awaiting review**
+* Impacted Platforms: [iOS / Android / Cloud]
+
+## Introduction
+
+Update the naïve menu updating with a new dynamic add / removal system of menu cells to use a more robust method that intelligently adds and removes menu cells, thereby cutting down on the amount of RPCs needed to update an SDL menu.
+
+## Motivation
+
+Currently, updating the menu through the menu manager in any way (even if the menu is only updated to add or remove a single cell) requires that the entire menu be deleted and added again. We should update our code to allow for updating the menu and dynamically determining which cells need to be deleted and which cells need to be added.
+
+## Proposed solution
+
+To solve this problem, we need a way to look at the old menu array, the new menu array, and determine adding / removing cells based on the differences.
+
+The first problem to be solved is to determine equality of cells. We don't currently do this for `SDLMenuCell`, but we do for `SDLChoiceCell`. The implementation of equality in `SDLMenuCell` will be very close to `SDLChoiceCell`, using the properties' values to create a hash and comparing those hashes.
+
+The second problem is to determine which adds and removals are necessary to turn the old / current menu array into the new menu array. We also need to consider sub-menus. Additionally complicating this issue is that existing menu cells cannot be moved, they must be removed and re-added, and if they have submenu cells, all those cells need to be removed and re-added as well.
```

Comment:
Change to:

The second problem is to determine which adds and removals are necessary to turn the old / current menu array into the new menu array. Sub-menus will also need to be considered. Additionally, complicating this issue is the existing menu cells that cannot be moved. They must be removed and re-added, and if they have submenu cells, all those cells need to be removed and re-added as well.

Reply:
Fixed